sdk/src/main/scala/com/horizen/account/state/AccountState.scala

```diff
@@ -0,0 +1,484 @@
+import com.horizen.account.utils.FeeUtils
+import com.horizen.account.block.AccountBlock
+import com.horizen.account.node.NodeAccountState
+import com.horizen.account.receipt.{EthereumReceipt, LogsBloom}
+import com.horizen.account.storage.AccountStateMetadataStorage
+import com.horizen.account.transaction.EthereumTransaction
+import com.horizen.account.utils.{AccountBlockFeeInfo, AccountFeePaymentsUtils, AccountPayment}
+import com.horizen.block.WithdrawalEpochCertificate
+import com.horizen.consensus.{ConsensusEpochInfo, ConsensusEpochNumber, ForgingStakeInfo, intToConsensusEpochNumber}
+import com.horizen.evm._
+import com.horizen.evm.interop.EvmLog
+import com.horizen.params.NetworkParams
+import com.horizen.state.State
+import com.horizen.utils.{ByteArrayWrapper, BytesUtils, ClosableResourceHandler, MerkleTree, TimeToEpochUtils, WithdrawalEpochInfo, WithdrawalEpochUtils}
+
+import org.web3j.crypto.ContractUtils.generateContractAddress
+import sparkz.core._
+import sparkz.core.transaction.state.TransactionValidation
+import sparkz.core.utils.NetworkTimeProvider
+import scorex.util.{ModifierId, ScorexLogging}
+import java.math.BigInteger
+import java.util
+import scala.collection.JavaConverters.seqAsJavaListConverter
+import scala.collection.mutable.ListBuffer
+import scala.util.{Failure, Success, Try}
+
+class AccountState(
+    val params: NetworkParams,
+    timeProvider: NetworkTimeProvider,
+    override val version: VersionTag,
+    stateMetadataStorage: AccountStateMetadataStorage,
+    stateDbStorage: Database,
+    messageProcessors: Seq[MessageProcessor]
+) extends State[SidechainTypes#SCAT, AccountBlock, AccountStateView, AccountState]
+      with TransactionValidation[SidechainTypes#SCAT]
+      with NodeAccountState
+      with ClosableResourceHandler
+      with ScorexLogging {
+
+  override type NVCT = AccountState
+
+  // Execute MessageProcessors initialization phase
+  // Used once on genesis AccountState creation
+  private def initProcessors(initialVersion: VersionTag): Try[AccountState] = Try {
+    using(getView) { view =>
+      for (processor <- messageProcessors) {
+        processor.init(view)
+      }
+      view.commit(initialVersion)
+      this
+    }
+  }
+
+  // Modifiers:
+  override def applyModifier(mod: AccountBlock): Try[AccountState] = Try {
+    require(
+      versionToBytes(version).sameElements(idToBytes(mod.parentId)),
+      s"Incorrect state version!: ${mod.parentId} found, " + s"$version expected"
+    )
+
+    using(getView) { stateView =>
+      if (stateView.hasCeased) {
+        val errMsg = s"Can't apply Block ${mod.id}, because the sidechain has ceased."
+        log.error(errMsg)
+        throw new IllegalStateException(errMsg)
+      }
+
+      // Check Txs semantic validity first
+      for (tx <- mod.sidechainTransactions)
+        tx.semanticValidity()
+
+      // TODO: keep McBlockRef validation in a view style, so in the applyMainchainBlockReferenceData method
```

Comment:
This TODO can be removed I think, we already are using `view.applyMainchainBlockReferenceData` below